krun/util.py

```diff
@@ -78,3 +83,65 @@ def run_shell_cmd(cmd, failure_fatal=True):
     if failure_fatal and rc != 0:
         fatal("Shell command failed: '%s'" % cmd)
     return stdout.strip(), stderr.strip(), rc
+
+
+def dump_results(config_file, out_file, all_results, audit):
+    """Dump results (and a few other bits) into a bzip2 json file."""
+    with open(config_file, "r") as f:
+        config_text = f.read()
+
+    to_write = {"config": config_text, "data": all_results, "audit": audit}
+
+    with bz2.BZ2File(out_file, "w") as f:
+        f.write(json.dumps(to_write, indent=1, sort_keys=True))
+
+
+def read_results(results_file):
+    results = None
+    with bz2.BZ2File(results_file, "rb") as f:
+        results = json.loads(f.read())
+    return results
+
+
+def check_and_parse_execution_results(stdout, stderr, rc):
+    json_exn = None
+    try:
+        iterations_results = json.loads(stdout)  # expect a list of floats
+    except Exception as e:  # docs don't say what can arise, play safe.
+        json_exn = e
+
+    if json_exn or rc != 0:
+        # Something went wrong
+        rule = 50 * "-"
+        err_s = ("Benchmark returned non-zero or didn't emit JSON list. ")
+        if json_exn:
+            err_s += "Exception string: %s\n" % str(e)
+        err_s += "return code: %d\n" % rc
+        err_s += "stdout:\n%s\n%s\n%s\n\n" % (rule, stdout, rule)
+        err_s += "stderr:\n%s\n%s\n%s\n" % (rule, stderr, rule)
+        raise ExecutionFailed(err_s)
+
+    return iterations_results
+
+
+def audits_same_platform(audit0, audit1):
+    """Check whether two platform audits are from identical machines.
+    A machine audit is a dictionary with the following keys:
+
+        * cpuinfo (Linux only)
+        * packages (Debian-based systems only)
+        * debian_version (Debian-based systems only)
+        * uname (all platforms)
+        * dmesg (all platforms)
+
+    Platform information may be Unicode.
```

Comment:
One way to do that would be to make this a method of the base class of platforms. However, I think platforms and results files are separate things, and I think we would be violating SRP if we do that. What I would suggest instead is to make a class that represents results files, which can compare two audits, read and write the results files, and so on. That way the execution scheduler would just receive a results file object as an argument to `build_schedule` and extract the previously executed keys via a method on the results file. 

This makes a lot more sense to me, in terms of OOP design and it allows for easier testing, but I think it needs to be a separate PR.